File: 2199-two-furthest-houses-with-different-colors/two-furthest-houses-with-different-colors.cpp
class Solution {
public:
    int maxDistance(vector<int>& colors) {
        int n = colors.size();     // total number of houses
        int maxDist = 0;           // max distance ko store karne ke liye variable

        // Case 1: Pehle house (index 0) ko last tak check karo 
        // aur dekho kaunsa aakhri house hai jo different color ka hai
        for (int i = n - 1; i >= 0; --i) {
            if (colors[i] != colors[0]) {
                maxDist = i - 0;   // distance calculate karo aur store karo
                break;             // jaise hi mil gaya, loop se bahar niklo
            }
        }

        // Case 2: Last house (index n-1) ko start se check karo 
        // aur dekho kaunsa pehla house hai jo different color ka hai
        for (int i = 0; i < n; ++i) {
            if (colors[i] != colors[n - 1]) {
                maxDist = max(maxDist, n - 1 - i);  // max distance update karo
                break;                              // jaise hi mil gaya, loop se bahar niklo
            }
        }

        return maxDist;  // final answer return karo
    }
};

File: 2199-two-furthest-houses-with-different-colors/two-furthest-houses-with-different-colors.py
class Solution:
    def maxDistance(self, colors: List[int]) -> int:
        n = len(colors)
        max_dist = 0

        # Case 1: first house se check
        for i in range(n - 1, -1, -1):
            if colors[i] != colors[0]:
                max_dist = i
                break

        # Case 2: last house se check
        for i in range(n):
            if colors[i] != colors[n - 1]:
                max_dist = max(max_dist, n - 1 - i)
                break

        return max_dist